### ISSUE TITLE
Documention of `do` is incorrect regarding error handling.

### ISSUE THREAD

**Issue by EvAlex** [ReactiveX/rxjs#2425]
There is inconsistency in how `do()` method handles errors. 

**RxJS version: 5.2**

**Code to reproduce:**
```js
Rx.Observable.of(1, 2, 3)
    .do(i => {
        if (i > 1) { 
            throw new Error(i + ' > 1') 
        } 
    })
    .subscribe(
        i => console.log(i), 
        e => console.error('Why? do() gave me different sequence. Initial one did not contain error. Initial one contained 3 values')
    );
```

**Expected behavior:**
Should produce the following console output:
```
1
Uncaught Error: 2 > 1
2
Uncaught Error: 3 > 1
3
```

**Actual behavior:**
Output is as follows:
```
1
Why? do() gave me different sequence. Initial one did not contain error. Initial one contained 3 values
```

**Additional information:**

[Documentation states that method do()](http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#instance-method-do):
> Perform a side effect for every emission on the source Observable, but return an Observable that is identical to the source.

but in next paragraph it states that

> Any errors that are thrown in the aforementioned Observer or handlers are safely sent down the error path of the output Observable.

Shouldn't it be considered as inconsistent statements? Initial observable can actually be modified in observer, passed in `do()` method. If that observer throws an error, the output observable will not be the same as initial. It will emit error, and it will not emit values after that error.

**Comment by staltz** [comment]
**TL;DR: we need to fix documentation. PRs welcomed.**

- - -

I think the problem here is in documentation 

> Perform a side effect for every emission on the source Observable, but return an Observable that **is identical to the source**.

We should say "identical to the source as long as errors don't occur" or "closely resembles the source".

The reason is, given
```js
const a$ = Rx.Observable.of(1, 2, 3);
const b$ = a$.do(i => {
        if (i > 1) { 
            throw new Error(i + ' > 1') 
        } 
    });
```

`a$` emits no errors. `b$` is an Observable that emits an error. This is because of:

> [do()] Any errors that are thrown in the aforementioned Observer or handlers are safely sent down the error path of the output Observable.

When b$ emits an error, b$ terminates. This is the typical observable contract. If an error goes to the `observer.error` handler, that means it is forbidden for `observer.next`/`observer.error`/`observer.complete` to be called after that.

**Comment by EvAlex** [comment]
Seems that the issue may be closed now.

**Comment by kwonoj** [comment]
Seems so. closed via #2543.
